frame/nomination-pools/src/lib.rs

```diff
@@ -1827,6 +1836,64 @@ pub mod pallet {
 
 			Ok(())
 		}
+
+		/// Update the roles of the pool.
+		///
+		/// The root is the only entity that can change any of the roles, including itself,
+		/// excluding the depositor, who can never change.
+		///
+		/// It emits an event, notifying UIs of the role change. This event is quite relevant to
+		/// most pool members and they should be informed of changes to pool roles.
+		#[pallet::weight(T::WeightInfo::update_roles())]
+		pub fn update_roles(
+			origin: OriginFor<T>,
+			pool_id: PoolId,
+			root: ConfigOp<T::AccountId>,
+			nominator: ConfigOp<T::AccountId>,
+			state_toggler: ConfigOp<T::AccountId>,
+		) -> DispatchResult {
+			let o1 = origin;
+			let o2 = o1.clone();
+
+			let is_pool_root = || -> Result<BondedPool<T>, sp_runtime::DispatchError> {
+				let who = ensure_signed(o1)?;
+				let bonded_pool = BondedPool::<T>::get(pool_id).ok_or(Error::<T>::PoolNotFound)?;
+				ensure!(bonded_pool.roles.root == who, Error::<T>::DoesNotHavePermission);
+				Ok(bonded_pool)
+			};
+			let is_root = || -> Result<BondedPool<T>, sp_runtime::DispatchError> {
+				ensure_root(o2)?;
+				let bonded_pool = BondedPool::<T>::get(pool_id).ok_or(Error::<T>::PoolNotFound)?;
+				Ok(bonded_pool)
+			};
+
+			let mut bonded_pool = is_root().or_else(|_| is_pool_root())?;
+
+			match root {
+				ConfigOp::Noop => (),
+				ConfigOp::Set(v) => bonded_pool.roles.root = v,
+				ConfigOp::Remove => Err(Error::<T>::CannotRemoveRole)?,
+			};
+			match nominator {
+				ConfigOp::Noop => (),
+				ConfigOp::Set(v) => bonded_pool.roles.nominator = v,
+				ConfigOp::Remove => Err(Error::<T>::CannotRemoveRole)?,
```

Comment:
I see that you want to use the `ConfigOp` here, but I think `Option<AccountId>` makes more sense for two reasons:  
1. Easier logic, just an `if let Some(nominator) = nominator`
2. Impossible to missuse, therefore `CannotRemoveRole` is not needed anymore since a `None` becomes a `Noop`.

Reply:
5e04c3d3804009bb44392eb59963ff129879190b